### PR TITLE
fix PodTopologySpread plugin conflict the volume node affinity

### LINKED ISSUE #116629: TopologySpreadConstraints - Scaling Pods during rolling updates can cause pods to be stuck in Pending
### What happened?

We have a stateful application that requires pods to be evenly spread across zones. Each pod required a persistent volume (AWS EBS volume) to store data. To deploy this application we use a StatefulSet with the following Pod Topology Spread Constraint:

```
topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            name: my-app
        maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: DoNotSchedule
```

During rolling updates, we update multiple pods from the same zone. For that, we use a custom controller ([zone-aware-controllers-for-k8s](https://github.com/aws/zone-aware-controllers-for-k8s)), but it would be possible to archive something similar using the [StatefulSets’ Max Unavailable](https://kubernetes.io/blog/2022/05/27/maxunavailable-for-statefulset/) alpha feature available on Kubernetes 1.24+.

When the application happens to be scaled (e.g manually or by the HPA) during a rolling updated, we have noticed old Pods getting stuck on Pending because of topology spread constraints + volume node affinity conflict: 

```
 Warning  FailedScheduling   40s   default-scheduler   0/34 nodes are available: 26 node(s) didn't match Pod's node affinity/selector, 4 node(s) didn't match pod topology spread constraints, 4 node(s) had volume node affinity conflict.
```

Example: 

Initial state (zone : pods).
- ZoneA : [PA1, PA2, PA3, PA4]
- ZoneB : [PB1, PB2, PB3, PB4]
- ZoneC : [PC1, PC2, PC3, PC4]

Assuming that we update 2 pods in Zone A:
- ZoneA : [PA1, PA2, **PA3(Terminating)**, **PA4(Terminating)**]
- ZoneB : [PB1, PB2, PB3, PB4]
- ZoneC : [PC1, PC2, PC3, PC4]

Now, scale the number of pods by the same number of pods getting updated. Noticed that new replicas are placed on Zone A first (scheduler seems to ignore the pods that are getting terminated/updated).
- ZoneA : [PA1, PA2, PA3(Terminating), PA4(Terminating), **PA5**, **PA6**]
- ZoneB : [PB1, PB2, PB3, PB4]
- ZoneC : [PC1, PC2, PC3, PC4]

In the end, some pods that were getting updated will be stuck in Pending as they can't be rescheduled into another zone because of the volume node affinity conflict (those pods can only be scheduled on ZoneA which is the zone in which their EBS volumes exists). 
- ZoneA : [PA1, PA2, PA3, **PA4(Pending)**, PA5, PA6]
- ZoneB : [PB1, PB2, PB3, PB4]
- ZoneC : [PC1, PC2, PC3, PC4]

Note that in this specific example PA3 will be scheduled on ZoneA as it satisfy `maxSkew = 1`, but PA4 will get stuck.

### What did you expect to happen?

Statefulset Pods getting restarted/updated are considered when scheduling new Pods.

### How can we reproduce it (as minimally and precisely as possible)?

The easier way to reproduce this is probably by deleting multiple pods from the same zone and, while new pods are getting recreated, try to scale the number of replicas. For example: 

- Create a Statefulset with PersistentVolume and the above TopologySpreadConstraint. 
- Scale the Statefulset so you have multiple pods per zone (e.g 12 pods distributed in 3 zones)
- Delete multiple pods from a single zone
- While pods are getting recreated, try to scale the number of replicas
- Notice that all (or majority) of new replicas are placed on the zone in which pods were deleted
- Notice that old pods are stuck in Pending

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"clean", BuildDate:"2022-01-25T21:17:57Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.16-eks-ffeb93d", GitCommit:"52e500d139bdef42fbc4540c357f0565c7867a81", GitTreeState:"clean", BuildDate:"2022-11-29T18:41:42Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
AWS EKS
</details>

### OS version

<details>

```console
N/A
```

</details>

### Install tools

<details>
N/A
</details>

### Container runtime (CRI) and version (if applicable)

<details>
N/A
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
[Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
</details>

### PR DESCRIPTION
**Pull request by olderTaoist**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
PodTopologySpread plugin will schedule new pod to the same node, when pod of statefulset with local storage volume(node affinity) was not exist,  will due to the old pod created can't schedule

#### Which issue(s) this PR fixes:
Fixes #116629

#### Does this PR introduce a user-facing change?

```release-note
Fixed PodTopologySpread plugin conflict the volume node affinity, which affect old not exist pod schedule
```

- [KEP]: [PodTopologySpread include terminated pod](https://github.com/kubernetes/enhancements/issues/4120)

### CODE REVIEW

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 1]
```diff
@@ -262,6 +264,61 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
...
+		var podReference metav1.OwnerReference
+		for _, reference := range pvc.OwnerReferences {
+			if reference.Kind == "StatefulSet" {
+				statefulSetReference = reference
+			}
+			if reference.Kind == "Pod" {
+				podReference = reference
+			}
+		}
+		node, err := pl.sharedLister.NodeInfos().Get(nodeName)
```
Comment (AxeZhan) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:289]: I think this line below can be placed in `filter`? We can store the reference message in state, since this logic is to handle the nodeInfo, I think place them in filter can reduce some delay between prefilter and filter. WDYT?

Also can we have some tests for this?

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 2]
```diff
@@ -262,6 +264,61 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
...
+		if err != nil {
+			return
+		}
+		var podExist bool
+		for _, podItem := range node.Pods {
+			if podItem.Pod.Namespace == pod.Namespace && podItem.Pod.Name == podReference.Name {
+				podExist = true
+				break
```
Comment (AxeZhan) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:297]: return in L3000 can be here?

Comment (olderTaoist) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: done

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 1 above)
Comment (olderTaoist) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: `calPreFilterState` calculate match for each topology pair depend on nodes、pods、constraints,  i think should place in `prefilter`.  i add unit test when you feel ready to lgtm.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 3]
```diff
@@ -19,11 +19,13 @@ package podtopologyspread
 import (
 	"context"
 	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (AxeZhan) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:22]: nit: this should line below: L25-L30

**Review by AxeZhan — COMMENTED** [review]
The implementation looks reasonable to me. Although I'm not sure if the scenario in the corresponding issue is worth this  implementation as there was some scenarios like this before #87621
You can now add some tests and wait for another reviewer.
/ok-to-test

File: pkg/controller/statefulset/stateful_set_utils.go [hunk 4]
```diff
@@ -342,6 +342,16 @@ func getPersistentVolumeClaims(set *apps.StatefulSet, pod *v1.Pod) map[string]v1
 		claim := templates[i]
 		claim.Name = getPersistentVolumeClaimName(set, &claim, ordinal)
 		claim.Namespace = set.Namespace
+		var ownReferences []metav1.OwnerReference
```
Comment (AxeZhan) [inline pkg/controller/statefulset/stateful_set_utils.go:345]:
```suggestion
		var ownerReferences []metav1.OwnerReference
```

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 5]
```diff
@@ -331,6 +358,60 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
...
+		if err != nil {
+			return
+		}
+		podItem := &v1.Pod{
+			Spec: sts.Spec.Template.Spec,
+		}
+		podInfo := &framework.PodInfo{
+			Pod: podItem,
+		}
+		podInfo.Update(podItem)
```
Comment (AxeZhan) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:404]: Why should we update the podInfo with sts's pod spec here?🤔️ I'm a little confused

**Comment by AxeZhan** [comment]
On a second thought after diving into https://github.com/kubernetes/kubernetes/issues/87621, will just **do not** by passing terminating Pods do it for this case?🤔️

**Comment by olderTaoist** [comment]
> On a second thought after diving into #87621, will just **do not** by passing terminating Pods do it for this case?🤔️

the issue discuss scheuduler whther exclude the teminating pod. here we include termintaed pod depend on volume with node affinity in `PodTopologySpread` plugin prefilter.  @alculquicondor @ahg-g PTAL

**Comment by AxeZhan** [comment]
> the issue discuss scheuduler whther exclude the teminating pod. here we include termintaed pod depend on volume with node affinity in `PodTopologySpread` plugin prefilter. 


Right, but in #116629, most time the target pod is in terminating/pending status right?

**Comment by alculquicondor** [comment]
Could you summarize what is this PR doing? That should generally be included in the PR description.

Overall, it doesn't seem like a clean design to make the topology spread plugin aware of PVCs. At least not with an accompanying KEP.

**Comment by olderTaoist** [comment]
> Could you summarize what is this PR doing? That should generally be included in the PR description.
> 
> Overall, it doesn't seem like a clean design to make the topology spread plugin aware of PVCs. At least not with an accompanying KEP.

please review the releated [KEP](https://github.com/kubernetes/enhancements/issues/4120) @alculquicondor

**Comment by alculquicondor** [comment]
Writing a KEP implies that you have to analyze any possible side-effects, or backwards-compatibility problems. I don't see that addressed.

**Comment by alculquicondor** [comment]
Why don't you start by describing your intended solution in the PR description?

**Comment by alculquicondor** [comment]
Unfortunately we are past the enhancements freeze.

We had too many open KEPs and we had no time to look at this one. For better outcome in the following release, I suggest you come to a SIG meeting to discuss the proposal.

**Comment by olderTaoist** [comment]
> Unfortunately we are past the enhancements freeze.
> 
> We had too many open KEPs and we had no time to look at this one. For better outcome in the following release, I suggest you come to a SIG meeting to discuss the proposal.

i already join [kubernetes-sig-scheduling](https://groups.google.com/forum/#!forum/kubernetes-sig-scheduling) group, will be participate next sig meeting
